### PR TITLE
add ``addons-linter`` package types

### PR DESCRIPTION
**Pull request by Robot-Inventor**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/addons-linter/addons-linter-tests.ts

Consider running `npx dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 86959f1ed63ad7954c62159e163aae2e0d4b8d57
</p>

File: types/addons-linter/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
+// Type definitions for addons-linter 6.15
```
Comment (jakebailey) [inline types/addons-linter/index.d.ts:1]: This package will need to be updated to the new layout.

**Review by jakebailey — CHANGES_REQUESTED** [review]

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/addons-linter/package.json

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4f3a1d5795fd0e12e2559f0d9d6e424bd0b04a61
</p>

File: types/addons-linter/index.d.ts (hunk 1 above)
Comment (Robot-Inventor) [reply types/addons-linter/index.d.ts]: Sorry for the inconvenience. I have updated to the new layout.

File: types/addons-linter/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    run: () => Promise<LinterResults>;
+}
+
+export function createInstance(options: Options): Linter;
+
+declare const _default: {
+    createInstance: typeof createInstance;
+};
+
+export default _default;
```
Comment (jakebailey) [inline types/addons-linter/index.d.ts:143]: I don't think this is right; the upstream library is CJS and its exported object has this shape:

```console
$ npm i "addons-linter"

added 252 packages, and audited 253 packages in 4s

found 0 vulnerabilities
$ node -e 'console.log(require("addons-linter"))'
{
  Linter: [class Linter],
  createInstance: [Function: createInstance],
  isRunFromCLI: [Function: isRunFromCLI]
}
```

So there doesn't seem to be a default.

Comment (Robot-Inventor) [reply types/addons-linter/index.d.ts]: It was my lack of confirmation.  Deleted default.

As a side note, as I added in the code comment, the Linter class and isRunFromCLI() function do not seem to be intended for external use, so I have not written a type definition for them.

Comment (jakebailey) [reply types/addons-linter/index.d.ts]: Yes, omitting stuff is totally fine.

File: types/addons-linter/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+        /**
+         * The log-level to generate.
+         */
+        logLevel: "fatal" | "error" | "warn" | "info" | "debug" | "trace";
+
+        /**
+         * Treat warnings as errors.
+         */
+        warningsAsErrors?: boolean;
```
Comment (jakebailey) [inline types/addons-linter/index.d.ts:16]: I highly suspect that these should all have `| undefined` to support `exactOptionalPropertyTypes`.

See: [Playground Link](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.2#code/PTAEHUEsBcAtQKYA8CGBjaB5ADtSB7AOxQBsAFAJ32wQugE8AVemgZ0WICMSEATAGlDMaAZTQVIuALAAoEKFgp2raBIy1QFAK492KKlsK9QAd0XRQ9fFtBoUhUKwQX7oaniKlZ87FRp16UGh8ADpZbzAAMXwKRFQAW2weQThIdkhCaFoAM3QEUF4ENBJ9BHY4FAs4Wny00BQI0F9qWgZTWEg0eDsHTnyifuygk3xHVQyAc1YALlAAcl59AGs5t1i5kkgJ2Gg5xpjQGEdYaxJjQnwLPsOHardOACsi6DCZWQysily0fIBVJwoABEELkdNB2ABvWSgUDyRiwfIoThOQg-NxDFCgABupC0+QoCF8ZQQmXYc1Y9BUCHiexkMLQ+BIMXh1IQmCxtAkhQA-LMAESLChLPmgAA+oD5m220D5AG5ZABfcJyMBQODWKoddjZEoTDhIngCIIIgn1U1wAn5HEkPHsMydeBWGw9RpOCwAAwZTIoLPibI5FC5CHdQVGfX5guFgklWx2IvsxndhkK2QyfHdr0aImceEIeru1rxodASaMILTvBDJkgJBI9Vr+BMoHi+BUoAAUigcWIJLhNIY8H7Gl0ikttQc7sg0llUflgqBcjXBPauod2CDss9IBySIFciQKa95AAJRsIAMpE0Icn1BdaABe97TrHKo1S7D6iixBAoISECIKctCBgAhCEabIDk9RlmQRP12U5SBChDAB9Xh8DKQg5mgZCbmKLRChLZMgPTTMVX-fIdRQPV3WQdAsFwUDSEoFoAmEMoQ3iFAljKIQWAQHtJBcNA0C0ChKgQXdGk46BVzuFMMhAogmiob9CmMJR6gcdxGLrZp-AYUjCmKUo71RDwHAmZx-lobNoFzKYAAoAEpZmsoEQRQMFWHlN4ZAZQg2zdez2AAXlASzoDc2zgucnygsmVgQi9GDWXgwNEPyMKBWWOVZHivNEuSn1YP9BCCKyqU4x8xoACEtAsC5oGmPKcwSpLoOK1KAyDUAwqI1NCD4aqyMoH9izuSjqNojAcHMpi-FaJg+NYd1LzqEwNMappW1YSBuAQEIgA)

Comment (Robot-Inventor) [reply types/addons-linter/index.d.ts]: Added ``| undefined`` to all optional properties.

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
@Robot-Inventor thanks for putting in the extra work to account for our monorepo upgrade.

**Pull request merged**